caddytls/ctmonitor.go

```diff
@@ -0,0 +1,308 @@
+package caddytls
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/mholt/caddy"
+)
+
+func init() {
+	caddy.RegisterEventHook("ctmonitor", startMonitoring)
+}
+
+const (
+	certSpotterAPIBase string = "https://api.certspotter.com/v1/issuances"
+)
+
+var (
+	filePath     string = filepath.Join(caddy.AssetsPath(), "ct_id")
+	ctConfigFile string = filepath.Join(caddy.AssetsPath(), "ct_config")
+)
+
+// CTConfig is the configuration for the CT monitor.
+type CTConfig struct {
+	Include_subdomains bool `json:"subdomains"`
+	Include_wildcards  bool `json:"wildcards"`
+}
+
+// CertSpotterResponse is the json response formatting that is used in the program.
+type CertSpotterResponse struct {
+	ID           string   `json:"id"`
+	TBSSHA256    string   `json:"tbs_sha256"`
+	DNSNames     []string `json:"dns_names"`
+	PubKeySha256 string   `json:"pubkey_sha256"`
+	Issuer       struct {
+		Name         string `json:"name"`
+		PubKeySha256 string `json:"pubkey_sha256"`
+	} `json:"issuer"`
+	NotBefore string `json:"not_before"`
+	NotAfter  string `json:"not_after"`
+	Cert      struct {
+		ID     string `json:"id"`
+		Type   string `json:"type"`
+		SHA256 string `json:"sha256"`
+		Data   string `json:"data"`
+	} `json:"cert"`
+}
+
+// QueryConfig is the configuration information for querying.
+type QueryConfig struct {
+	CTConfig
+	Query string
+	Index int
+}
+
+// CompareCerts compares the certificates that caddy is serving against the certificates
+// that certSpotter has found, if there are any that don't match the caddy certificates,
+// they are reported to the user.
+func CompareCerts(caddyCerts map[string]struct{}, certSpotterCerts map[string]string) {
+	for key := range certSpotterCerts {
+		if _, ok := caddyCerts[key]; !ok {
+			log.Printf("[WARNING] Certificate found that caddy is not monitoring, issued by: %v\n", certSpotterCerts[key])
```

Comment:
@JosephSummerhays - we need to add an entry to telemetry here in addition to the log entry.

Reply:
@mholt I added the line to the program, but telemetry wasn't recognized, am I supposed to include a telemetry header, or tell it where to find the telemetry tools inside of caddy?

Reply:
I got the telemetry package imported properly and the line for the telemetry added.